### PR TITLE
React-Feature-Toggles

### LINKED ISSUE #105: React-Feature-Toggles
* [x] **Add react-feature-toggles to package**
* [x] **May need to update React to allow for new context api?**
* [x] **May need to update @types/react to allow for new context api?**

### PR DESCRIPTION
**Pull request by jwicks31**
Resolves #105 

### Description of Changes

Installing react-feature-toggles required updates to a few dependencies.

- Updates `react` & `react-dom` to `v16.4.1`
- Updates `@types/react` to `v16.3.17`
- Updates `react-hot-loader` to `v3.1.3`
       -This is not the most up to date version, but v1 would not work with React 16. The current version does not support webpack.
- Installs react-feature-toggles package
- Adds FeatureToggle provider component from `react-feature-toggles`
- Adds AppContainer provider component from `react-hot-loader `
- Adds `react-feature-toggles` typing declaration `.d.ts` file.

**TO TEST REACT-FEATURE-TOGGLES**
Open Netlify Link
screen shot 2018-06-20 at 4 44 56 pm

Add '?ft=foo' to the end of the url. 
screen shot 2018-06-20 at 4 44 48 pm

### PR Review Checklist

* [ ] **Have you tested the changes?**
* [ ] **Is QE review required? If so, has this been reviewed by QE?** (For all new features, review by QE is required. Minor changes which already have passing functional tests may not need a new QE review -- when in doubt, loop in a QE reviewer).
* [ ] **Are there code changes that require new unit tests? If so, are those tests included?** (A separate PR for tests is not appropriate).
* [ ] **Are documentation updates required? If so, does this PR include it?** (Separate documentation PR is generally not appropriate).
* [ ] **Are the added/edited files named appropriately and located correctly in the project directory structure?**

### CODE REVIEW

File: src/typings/react-feature-toggles/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  interface Query {
+    readonly ft: string
+  }
+  function getQueryFeatures(query?: Query): ReadonlyArray<string>
+  export { getQueryFeatures }
+
+  function mergeFeatures(
+    currentFeatures: ReadonlyArray<string>,
+    restOfFeatures: ReadonlyArray<string>
```
Comment (jwicks31) [inline src/typings/react-feature-toggles/index.d.ts]: restOfFeatures should be:
`...restOfFeatures: ReadonlyArray<string>`
I am not sure how to do the rest operator.

File: src/typings/react-feature-toggles/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  export { isActive }
+
+  interface Req {
+    readonly query: Query
+  }
+
+  function getReqQueryFeatures(req?: Req): ReadonlyArray<string>
+  export { getReqQueryFeatures }
+
+  function getBrowserQueryFeatures(search?: string): ReadonlyArray<string>
```
Comment (jwicks31) [inline src/typings/react-feature-toggles/index.d.ts]: This was not working for some reason. I am having trouble with it.

File: src/components/atoms/Hash/Hash.stories.tsx [hunk 3]
```diff
@@ -16,7 +16,7 @@ storiesOf('Components/Text', module).addWithJSX(
   ### Usage
   ~~~js
   <Hash
-    readonly children?: JSX.Element
+    readonly children?: any
```
[the remark this answers is not in the document]
Comment (jwicks31) [reply src/components/atoms/Hash/Hash.stories.tsx]: It was not building after I had updated React. At first some things I read where saying that the recommendation was to use `children?: any`, but I read a little more over the weekend and I believe React.ReactNode is the correct type for most of them. There are some that could possibly just be strings? But I wasn't sure how to handle them Ex. Hash and HeaderTitle.

File: src/typings/react-feature-toggles/index.d.ts (hunk 1 above)
Comment (lautarodragan) [reply src/typings/react-feature-toggles/index.d.ts]: See https://github.com/Microsoft/TypeScript/issues/15972

Comment (jwicks31) [reply src/typings/react-feature-toggles/index.d.ts]: Ok cool! That seems to have worked! I noticed that the ReadonlyArray wasn't an option, just wasn't able to figure it out/find the answer after that. Thank you!

Comment (jwicks31) [reply src/typings/react-feature-toggles/index.d.ts]: Well, it worked on my local machine, but out lint rule changes it to ReadonlyArray<string>.

File: src/typings/react-feature-toggles/index.d.ts (hunk 2 above)
Comment (lautarodragan) [reply src/typings/react-feature-toggles/index.d.ts]: What problem are you having with this one?

Comment (jwicks31) [reply src/typings/react-feature-toggles/index.d.ts]: I'm honestly not really sure what is going on. I thought that I had the types correct and everything, but whenever I'm trying to test to see if it actually works it says that it isn't a function. ```TypeError: react_feature_toggles_1.getBrowserQueryFeatures is not a function```

Comment (jwicks31) [reply src/typings/react-feature-toggles/index.d.ts]: This was a problem with react-feature-toggles, I fixed it over there.

**Comment by geoffturk** [comment]
Nice!

image

🎆

File: src/typings/react-feature-toggles/index.d.ts (hunk 1 above)
Comment (wzalazar) [reply src/typings/react-feature-toggles/index.d.ts]: I think it's possible to avoid this rule in this line if you need

Comment (jwicks31) [reply src/typings/react-feature-toggles/index.d.ts]: I updated to `...restOfFeatures: string[]` with the rule, but it is giving this error: ```Argument of type 'string[]' is not assignable to parameter of type 'string'.``` Trying to work that out before I merge.

Comment (jwicks31) [reply src/typings/react-feature-toggles/index.d.ts]: I think the correct solution was `...restOfFeatures: Array<ReadonlyArray<string>>`

Comment (wzalazar) [reply src/typings/react-feature-toggles/index.d.ts]: LGTM @lautarodragan do you agree?

**Pull request merged**